DIPs/DIP1NNN-DK.md

```diff
@@ -0,0 +1,617 @@
+    U u;
+    *u.asPtr = 3; // error: field U.asPtr cannot access pointers in @safe code that overlap other fields
+}
+```
+
+However, sometimes using these operations is necessary for performance.
+Take for example a [tagged union](https://en.wikipedia.org/wiki/Tagged_union).
+To save on memory, it overlaps multiple fields of possibly different types, and then uses an integer to keep track which type of the union is currently valid.
+As discussed above however, this is not allowed in `@safe` code when pointers, arrays or classes are among the union members.
+One idea is to encapsulate a tagged union using a templated type and use `@trusted` getter methods that ensure the pointer members are only accessed when the tag ensures the underlying data is a valid pointer.
+The package [sumtype](http://code.dlang.org/packages/sumtype) does this.
+
+There is a loophole however: By meddling with the `tag` field, memory can be corrupted in `@safe` code:
+
+```D
+/+ dub.sdl:
+dependency "sumtype" version="~>0.8.11"
++/
+import sumtype;
+import std;
+
+void main() @safe {
+	immutable(int)* immutPtr = new int(10);
+
+ 	SumType!(immutable(int)*, int*) sum = immutPtr; // assign immutable pointer
+
+	__traits(getMember, sum, "tag") = 1; // ! ruin integrity of tagged union
+
+	sum.match!(
+		(ref immutable(int)* ptr) {}, // this one should be called
+		(ref int* ptr) {*ptr = 1000;}, // but we made it call this one
+	);
+
+	writeln(*immutPtr); // prints '1000'. the immutable pointer got mutated!
+}
+```
+
+It actually isn't possible to fix this currently, because:
+- the tag having `private` visibility only encapsulates on the module level, while `@trusted` is applied on the function level, so functions in the same module can still violate it
+- even outside the module, `__traits(getMember, )` [bypasses `private`](https://github.com/dlang/dmd/pull/9585)
+
+A way of restricting access to the `tag` field from `@safe` code is needed to allow a `@safe` semantic type like this.
+The tagged union is just one example, there are many other situations where data should only be modified by code that "knows what it's doing".
+This is especially necessary for the recent push of `@safe` non-garbage collected data structures / system programming as evidenced by:
+- [Ownsership and borrowing in D](https://dlang.org/blog/2019/07/15/ownership-and-borrowing-in-d/)
+- [DIP1021](https://github.com/dlang/DIPs/blob/master/DIPs/accepted/DIP1021.md)
+- [my vision of D's future](https://dlang.org/blog/2019/10/15/my-vision-of-ds-future/)
+
+Take reference counting for example: As long as the reference count can be meddled with from `@safe` code, freeing the memory can't be `@trusted` and D won't support `@safe` reference counting.
+
+### Existing holes in `@safe`
+
+While constructing arbitrary pointers inside `@safe` code is not allowed, there is currently no defence around pointers initialized to an unsafe value outside any function:
+```D
+@safe:
+int* x = cast(int*) 0x7FFE_E800_0000;
```

Comment:
This doesn't compile anymore with DMD 2.088. https://github.com/dlang/dmd/pull/10056

Reply:
What a surprise! It still compiles when marking the declarations individually with `@safe`, and I assumed putting `@safe:` above it would be identical.
The fix seems to only be for unsafe casts, and the following still compiles:
```D
auto getPtr() @system {return cast(int*) 0x7FFE_E800_0000;}

@safe:
int* x = getPtr();

void main() @safe {
    *x = 3; // memory corruption
}
```
So I'll update the example.